pkg/server/singleprocess/service_pipeline.go

```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+				// is then that's an internal error on the caller
+				return nil, nil, status.Errorf(codes.Internal,
+					"parentStep cannot be empty string if building an embedded graph for pipeline %q!",
+					pipeline.Name)
+			}
+
+			// Add an edge to the parent step as an implicit dependency
+			// Embedded pipeline steps have an implicit dependency on the parent step
+			// from the parent pipeline.
+			stepGraph.AddEdge(parentStep, nodeId)
+		}
+
+		// Add any dependencies as defined by the current Step
+		for _, dep := range step.DependsOn {
+			// Look up the dependency step by name in case we've already generated
+			// a node ID for it for our stepGraph
+			depId, ok := s.stepToNodeId(ctx, log, pipeline.Id, dep, nodeIdMap)
+			if !ok {
+				var err error
+				// We haven't reached this node yet, but we're adding it to the graph so
+				// we generate an id here too so we can add it to the graph as a vertex
+				// and create an edge to the given step with the depencny
+				depId, err = server.Id() // unique node graph id for full graph
+				if err != nil {
+					return nil, nil, err
+				}
+
+				// add node id to map
+				nodeIdMap[depId] = &pb.Ref_PipelineStep{
+					Pipeline: pipeline.Id,
+					Step:     dep,
+				}
+			}
+
+			// Add the dependency as a vertex and draw an edge to *this* steps node ID
+			stepGraph.Add(depId)
+			stepGraph.AddEdge(depId, nodeId)
+
+			// This only checks for steps inside *this* pipeline. Plain steps cannot
+			// reference other steps in other pipelines as a depenency without being
+			// an embedded pipeline reference.
+			if _, ok := pipeline.Steps[dep]; !ok {
+				return nil, nil, fmt.Errorf(
+					"step %q depends on non-existent step %q", step, dep)
+			}
+		}
+
+		// Build a graph for any nested pipelines
+		if embedRef, ok := step.Kind.(*pb.Pipeline_Step_Pipeline_); ok {
+			// This is only a "ref" to the pipeline, we have to
+			// look it up here to get the actual steps.
+			embeddedPipeline, err := s.state(ctx).PipelineGet(embedRef.Pipeline.Ref)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			// Build the nested pipelines graph
+			parentStep := nodeId
+			embeddedGraph, n, err := s.pipelineGraphFull(ctx, log, stepGraph,
+				parentStep, visistedNodes, nodeIdMap, embeddedPipeline)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			nodeIdMap = n
+			stepGraph = embeddedGraph
 		}
 	}
 
-	run.State = pb.PipelineRun_STARTING
-	if err = s.state(ctx).PipelineRunPut(run); err != nil {
-		return nil, err
+	if cycles := stepGraph.Cycles(); len(cycles) > 0 {
+		return nil, nil, status.Errorf(codes.FailedPrecondition,
+			"step dependencies in pipeline %q contain one or more cycles: %s",
+			pipeline.Name, cycles)
 	}
 
-	// Queue all the jobs atomically
-	if _, err := s.queueJobMulti(ctx, stepJobs); err != nil {
-		return nil, err
+	return stepGraph, nodeIdMap, nil
+}
+
+func (s *Service) stepToNodeId(
+	ctx context.Context,
+	log hclog.Logger,
+	pipelineName string,
```

Comment:
```suggestion
	pipelineId string,
```

`stepRef.Pipeline` is the ID of the pipeline, but the variable is `pipelineName`, but this method is always called with the input variable `pipeline.Id`. So we should rename this variable to be clearer.

_OR_ if we want to always reference the pipeline by Name and not ID, then we should refactor `Ref_PiplineStep`.

Reply:
Makes sense to me to rename this local var `pipelineId`!